wasm-chain-integration/src/lib.rs

```diff
@@ -337,12 +371,66 @@ impl<'a> HasCommon for ReceiveHost<'a> {
 
     fn param(&self) -> &[u8] { &self.param }
 
-    fn metadata(&self) -> &ChainMetadata { &self.receive_ctx.metadata }
+    fn metadata(&self) -> ExecResult<&Self::MetadataType> { self.receive_ctx.metadata() }
+
+    fn policies(&self) -> ExecResult<&P> { self.receive_ctx.sender_policies() }
+}
+
+pub trait HasInitContext<Policies = Vec<OwnedPolicy>> {
+    type MetadataType: HasChainMetadata;
+    fn metadata(&self) -> ExecResult<&Self::MetadataType>;
+    fn init_origin(&self) -> ExecResult<AccountAddress>;
+    fn sender_policies(&self) -> ExecResult<&Policies>;
+}
+
+impl<Policies> HasInitContext<Policies> for InitContext<Policies> {
+    type MetadataType = ChainMetadata;
+
+    fn metadata(&self) -> ExecResult<&Self::MetadataType> { Ok(&self.metadata) }
+
+    fn init_origin(&self) -> ExecResult<AccountAddress> { Ok(self.init_origin) }
+
+    fn sender_policies(&self) -> ExecResult<&Policies> { Ok(&self.sender_policies) }
+}
+
+pub trait HasReceiveContext<Policies = Vec<OwnedPolicy>> {
+    type MetadataType: HasChainMetadata;
+    fn metadata(&self) -> ExecResult<&Self::MetadataType>;
```

Comment:
Shouldn't this always return a metadatatype, which then can have optional fields?

Is there value in this being optional as well?